src/v/storage/disk_log_appender.cc

```diff
@@ -61,6 +61,9 @@ bool disk_log_appender::needs_to_roll_log(model::term_id batch_term) const {
      * _bytes_left_in_segment is for initial condition
      *
      */
+    if (!_seg || !_seg->has_appender()) {
+        return true;
+    }
```

Comment:
question: Why is this conditional required? Reading it, I expected us to return `false` here. If the appender was released concurrently (which I think is what `!_seg->has_appender()` is targeting), shouldn't we skip the roll?

Reply:
Specific races don't come to mind since this is all done under the rolling lock, but this line is meant to answer "why is it safe for `append_batch_to_segment()` to call `_seg->append()`?"

Reply:
Even though we're holding this lock, it's possible segment.ms previously released the appender of the current segment. In that case, the `maybe_roll_unlocked()` wouldn't create a new segment, and we'd proceed to `initialize()` and then append safely.

Reply:
I renamed some of this stuff so hopefully it's clearer.